### PR TITLE
fix: token authentication for API requests 

### PR DESCRIPTION
**Pull request by kjappelbaum**
all `PUT` requests failed with unauthorized as the `getUserEmailFromToken` function always returned `anonymous`.  Now it checks for the owner of the token.

Test cases for requests using the tokens have been added.

### CODE REVIEW

**Comment by kjappelbaum** [comment]
@lpatiny helped me debugging this by sharing the logs from eln.epfl.ch which had 

```
rest-on-couch_1            | 2021-03-02T08:35:16.122Z couch:debug (main:token) getToken (taen8ECx9GvtiTyyCKxkoOTsCgxdGGXF)
rest-on-couch_1            | 2021-03-02T08:35:16.139Z couch:debug (main:attachment) addAttachments (0a44cb583a15781f6c0ea5071b63e9bd, anonymous)
rest-on-couch_1            | 2021-03-02T08:35:16.140Z couch:debug (main:entry) getEntryWithRights (0a44cb583a15781f6c0ea5071b63e9bd, anonymous, [ 'write', 'addAttachment' ])
rest-on-couch_1            | 2021-03-02T08:35:16.182Z couch:warn (main:nano) Getting more than one result for a group name
rest-on-couch_1            | 2021-03-02T08:35:16.185Z couch:warn (main:nano) Getting more than one result for a group name
nginx-proxy_1              | 172.18.0.1 - - [02/Mar/2021:08:35:16 +0000] "PUT /roc/db/eln/entry/0a44cb583a15781f6c0ea5071b63e9bd/spectra/xray/05000N3.cif?token=taen8ECx9GvtiTyyCKxkoOTsCgxdGGXF HTTP/1.1" 401 52 "-" "curl/7.29.0" "-
```

for the `PUT` requests we made (to some sample we owner of the token owned) using a token with `addAttachment` rights.

File: src/server/middleware/auth.js [hunk 1]
```diff
@@ -97,30 +97,12 @@ exports.getProvider = function (ctx) {
...
-          cookie: token,
-        },
-        throwHttpErrors: false,
-      },
-    );
-
-    if (res.body && res.body.userCtx) {
-      return res.body.userCtx.name;
-    }
+  try {
```
Comment (targos) [inline src/server/middleware/auth.js:100]: this function is probably poorly name, because it has nothing to do with `?token=` authentication.

Comment (kjappelbaum) [reply src/server/middleware/auth.js:100]: yes, this was something that confused me. I was not sure what this was supposed to do.

Comment (kjappelbaum) [reply src/server/middleware/auth.js:100]: right, but we can still leave the the previous code in for the `x-auth-session` and then have this check for the `?token=` case?

Comment (kjappelbaum) [reply src/server/middleware/auth.js:100]: that is, something like---but again, sorry if i'm wasting your time here!

```js
async function getUserEmailFromAuthToken(ctx) {
  try {
    const token = await ctx.state.couch.getToken(ctx.query.token);
    return token.$owner;
  } catch (err) {
    return 'anonymous';
  }
}
async function getUserEmailFromToken(ctx) {
  // Try to get ?token token and use this and not the x-auth cookie token
  const authToken = getUserEmailFromAuthToken(ctx);
  if (authToken !== 'anonymous') return authToken;

  if (!config.authServers.length) return 'anonymous';

  const token = ctx.headers['x-auth-session'] || ctx.query['x-auth-session'];
  if (!token) return 'anonymous';

  for (let i = 0; i < config.authServers.length; i++) {
    // eslint-disable-next-line no-await-in-loop
    const res = await got(
      `${config.authServers[i].replace(/\/$/, '')}/_session`,
      {
        responseType: 'json',
        headers: {
          cookie: token,
        },
        throwHttpErrors: false,
      },
    );

    if (res.body && res.body.userCtx) {
      return res.body.userCtx.name;
    }
  }

  return 'anonymous';
}
```

File: test/data/data.js [hunk 2]
```diff
@@ -88,6 +97,30 @@ function populate(db) {
...
+
+  prom.push(
+    insertDocument(db, {
+      $type: 'token',
+      $kind: 'entry',
+      $owner: 'b@b.com',
+      $id: 'myAddAttachmentToken',
+      $creationDate: 0,
+      uuid: 'A',
+      rights: ['read', 'addAttachment'],
```
Comment (targos) [inline test/data/data.js:120]:
```suggestion
      rights: ['read', 'write', 'addAttachment'],
```

The "write" right is necessary to upload an attachment

File: src/server/middleware/auth.js (hunk 1 above)
Comment (targos) [reply src/server/middleware/auth.js:100]: Please revert this part of the change. The bug is not here and it's not how it should be fixed. An "entry" token only gives the right to a specific entry. If you override the user email with the one that's in the token, you effectively allow that request to access everything that the user can access (not only the entry specified by the token, and also not only with the rights specified in the token).

File: src/server/middleware/couch.js [hunk 3]
```diff
@@ -18,18 +18,21 @@ const couchNeedsParse = ['key', 'startkey', 'endkey'];
 const invalidDbName = 'invalid database name';
 exports.setupCouch = async (ctx, next) => {
   const dbname = ctx.params.dbname;
+  debug('setting up couch with db', dbname);
```
Comment (targos) [inline src/server/middleware/couch.js:21]:
```suggestion
  debug.trace('setting up couch with db', dbname);
```

File: src/server/middleware/couch.js [hunk 4]
```diff
@@ -18,18 +18,21 @@ const couchNeedsParse = ['key', 'startkey', 'endkey'];
...
+    debug('error setting up couch', e);
     if (e.message === invalidDbName) {
       onGetError(ctx, new CouchError(invalidDbName, 'forbidden'));
       return;
     }
   }
+  ctx.state.userEmail = ctx.query.asAnonymous
+    ? 'anonymous'
+    : await auth.getUserEmail(ctx);
+  debug('user email set to', ctx.state.userEmail);
```
Comment (targos) [inline src/server/middleware/couch.js:35]:
```suggestion
  debug.trace('user email set to', ctx.state.userEmail);
```

File: test/data/data.js [hunk 5]
```diff
@@ -29,6 +29,15 @@ function populate(db) {
...
   );
 
+  prom.push(
+    insertDocument(db, {
+      $type: 'entry',
+      $owners: ['a@a.com'],
+      $id: 'documentOfA',
+      $content: {},
+    }),
+  );
```
Comment (targos) [inline test/data/data.js:39]: you need to specify the `_id` corresponding to the token's `uuid` to make the test work correctly

File: test/data/data.js [hunk 6]
```diff
@@ -88,6 +97,30 @@ function populate(db) {
...
 
+  prom.push(
+    insertDocument(db, {
+      $type: 'token',
+      $kind: 'entry',
+      $owner: 'b@b.com',
+      $id: 'myReadOnlyToken',
+      $creationDate: 0,
+      uuid: 'A',
+      rights: ['read'],
```
Comment (targos) [inline test/data/data.js:108]: to be clear, this token should only allow to "read" the entry with uuid "A". any other action on that entry and any action on other entries should not be allowed (unless the entry is public)

File: test/unit/rest-api/user.js [hunk 7]
```diff
@@ -13,6 +13,31 @@ describe('User REST-api (data, anonymous)', () => {
...
+      .send('rest-on-couch!!')
+      .expect(200);
+  });
+
+  test('Should be able to PUT but not without the right in the token', () => {
+    return request
+      .put('/db/test/entry/A/blub.txt?token=myReadOnlyToken')
+      .set('Content-Type', 'text/plain')
+      .send('rest-on-couch!!')
+      .expect(200);
```
Comment (targos) [inline test/unit/rest-api/user.js:31]: this request should not be allowed

File: test/unit/rest-api/user.js [hunk 8]
```diff
@@ -13,6 +13,31 @@ describe('User REST-api (data, anonymous)', () => {
...
+  });
+
+  test('Should be able to GET with the token', () => {
+    return request
+      .get('/db/test/entry/A?token=myAddAttachmentToken')
+      .expect(200);
+  });
+});
```
Comment (targos) [inline test/unit/rest-api/user.js:39]: you can also add a similar check with the read-only token

**Comment by targos** [comment]
The two places to fix are:
https://github.com/cheminfo/rest-on-couch/blob/e348e7ffdcb6c58f75322d4b76b31eab2a5812db/src/server/middleware/couch.js#L119-L137
https://github.com/cheminfo/rest-on-couch/blob/e348e7ffdcb6c58f75322d4b76b31eab2a5812db/src/couch/attachment.js#L11-L43

What we need to do is pass `ctx.query.token` to the options of `getEntryWithRights`. You can see how it's done already in other methods such as [`getDocument`](https://github.com/cheminfo/rest-on-couch/blob/e348e7ffdcb6c58f75322d4b76b31eab2a5812db/src/server/middleware/couch.js#L74-L80)

File: src/server/middleware/auth.js (hunk 1 above)
Comment (targos) [reply src/server/middleware/auth.js:100]: The token lookup is already done for all requests at:
https://github.com/cheminfo/rest-on-couch/blob/master/src/server/routes/api.js#L21

**Comment by kjappelbaum** [comment]
Thanks so much for your feedback! I'm learning a lot and hope it is not too annoying it i "fixed" the bug in a completely wrong place ;) Will now just add some tests for the readonly token

File: src/couch/entry.js [hunk 9]
```diff
@@ -11,7 +11,7 @@ const validateMethods = require('./validate');
 
 const methods = {
   async getEntryWithRights(uuid, user, rights, options = {}) {
-    debug('getEntryWithRights (%s, %s, %o)', uuid, user, rights);
+    debug('getEntryWithRights (%s, %s, %o, %o)', uuid, user, rights, options);
```
Comment (targos) [inline src/couch/entry.js:14]:
```suggestion
    debug('getEntryWithRights (%s, %s, %o)', uuid, user, rights);
```

Printing the entire token is too much for a regular debug

**Comment by targos** [comment]
No problem, I know this is not an easy codebase :)

**Review by targos — APPROVED** [review]

**Pull request merged**